File: android/app/src/main/java/com/transitkit/app/data/push/PushApiClient.kt
package com.transitkit.app.data.push

import android.os.Build
import com.squareup.moshi.JsonClass
import com.squareup.moshi.Moshi
import com.transitkit.app.config.OperatorConfig
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Talks to the transitkit-console CMS for public endpoints — currently just
 * the test-device registration. The base URL comes from
 * `OperatorConfig.consoleApiUrl`. In debug builds we override to
 * `http://10.0.2.2:3000` (the Android emulator's loopback to the host) so
 * dev smoke tests against a locally running CMS work out of the box.
 */
@Singleton
class PushApiClient @Inject constructor(
    private val config: OperatorConfig,
) {
    private val http = OkHttpClient()
    private val moshi = Moshi.Builder().build()
    private val adapter = moshi.adapter(RegisterRequest::class.java)

    @JsonClass(generateAdapter = true)
    data class RegisterRequest(
        val operator_id: String,
        val fcm_token: String,
        val platform: String,
        val label: String,
    )

    sealed class Result {
        data class Ok(val id: String?) : Result()
        data class Err(val message: String) : Result()
    }

    /** Resolve the CMS base URL.
     *  - Debug build:   `http://10.0.2.2:3000` (emulator → host loopback)
     *  - Release build: the operator's `console_api_url` from config.json */
    fun resolveBaseUrl(): String? {
        if (com.transitkit.app.BuildConfig.DEBUG) {
            return "http://10.0.2.2:3000"
        }
        return config.consoleApiUrl?.takeIf { it.isNotBlank() }
    }

    fun defaultDeviceLabel(): String {
        val manufacturer = Build.MANUFACTURER.replaceFirstChar { it.uppercase() }
        val model = Build.MODEL
        val display = if (model.startsWith(manufacturer, ignoreCase = true)) model
                      else "$manufacturer $model"
        return "$display — Android ${Build.VERSION.RELEASE}"
    }

    suspend fun registerTestDevice(token: String, label: String): Result {
        val base = resolveBaseUrl()?.trimEnd('/')
            ?: return Result.Err("console_api_url missing from operator config")
        val url = "$base/api/test-devices/register"

        val body = adapter.toJson(
            RegisterRequest(
                operator_id = config.id,
                fcm_token = token,
                platform = "android",
                label = label,
            )
        ).toRequestBody("application/json".toMediaType())

        return try {
            val response = http.newCall(
                Request.Builder().url(url).post(body).build()
            ).execute()
            response.use { resp ->
                if (resp.isSuccessful) {
                    Result.Ok(resp.body?.string())
                } else {
                    Result.Err("HTTP ${resp.code}: ${resp.body?.string()?.take(200) ?: ""}")
                }
            }
        } catch (e: Exception) {
            Result.Err("Network error: ${e.message}")
        }
    }
}

File: ios/TransitKit/Sources/Services/PushApiClient.swift
import Foundation
import UIKit

// MARK: - PushApiClient

/// Talks to the transitkit-console CMS for endpoints that are public
/// (no Supabase JWT) — currently just the test-device registration.
///
/// The base URL comes from `OperatorConfig.consoleApiUrl`. When the field
/// is missing in config.json the client refuses to send.
enum PushApiClient {
    /// Resolve the CMS base URL.
    ///
    /// - Debug build:   `TRANSITKIT_LOCAL_CONSOLE` env → else `http://localhost:3000`.
    ///                  The operator's production URL is ignored so the dev
    ///                  loop never accidentally hits prod.
    /// - Release build: the operator's `console_api_url` from config.json.
    ///
    /// The Info.plist `NSAllowsLocalNetworking` exception covers the http://
    /// localhost case for Debug.
    static func resolveBaseUrl(operatorConsoleUrl: String?) -> String? {
        #if DEBUG
        if let envOverride = ProcessInfo.processInfo.environment["TRANSITKIT_LOCAL_CONSOLE"],
           !envOverride.isEmpty {
            return envOverride
        }
        return "http://localhost:3000"
        #else
        return operatorConsoleUrl
        #endif
    }

    struct RegisterTestDeviceRequest: Encodable {
        let operator_id: String
        let fcm_token: String
        let platform: String
        let label: String
    }

    enum Failure: LocalizedError {
        case missingBaseUrl
        case invalidUrl
        case httpError(status: Int, body: String?)
        case transport(Error)
        case decoding(Error)

        var errorDescription: String? {
            switch self {
            case .missingBaseUrl: "console_api_url missing from operator config.json"
            case .invalidUrl:     "console_api_url is not a valid URL"
            case let .httpError(status, body): "Server replied \(status)\(body.map { ": \($0)" } ?? "")"
            case let .transport(e): "Network error: \(e.localizedDescription)"
            case let .decoding(e):  "Decoding error: \(e.localizedDescription)"
            }
        }
    }

    static func registerTestDevice(
        consoleBaseUrl: String?,
        operatorId: String,
        fcmToken: String,
        label: String
    ) async throws {
        let resolved = resolveBaseUrl(operatorConsoleUrl: consoleBaseUrl)
        guard let base = resolved, !base.isEmpty else { throw Failure.missingBaseUrl }
        guard let url  = URL(string: base.trimmingCharacters(in: .init(charactersIn: "/")) + "/api/test-devices/register") else {
            throw Failure.invalidUrl
        }

        let body = RegisterTestDeviceRequest(
            operator_id: operatorId,
            fcm_token:   fcmToken,
            platform:    "ios",
            label:       label
        )

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw Failure.transport(URLError(.badServerResponse))
            }
            if !(200..<300).contains(http.statusCode) {
                let bodyText = String(data: data, encoding: .utf8)
                throw Failure.httpError(status: http.statusCode, body: bodyText)
            }
        } catch let failure as Failure {
            throw failure
        } catch {
            throw Failure.transport(error)
        }
    }

    /// A user-facing label like "iPhone 16 Pro (sim)" / "iPhone di Marco".
    @MainActor
    static func defaultDeviceLabel() -> String {
        let name = UIDevice.current.name
        let model = UIDevice.current.model
        // `UIDevice.name` returns generic "iPhone" on iOS 16+ unless the
        // app has the special entitlement. Show model as a fallback.
        if name.localizedCaseInsensitiveContains("iPhone") || name == model {
            return "\(model) — \(UIDevice.current.systemName) \(UIDevice.current.systemVersion)"
        }
        return name
    }
}
